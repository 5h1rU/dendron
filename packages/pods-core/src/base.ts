import { DEngine } from "@dendronhq/common-all";
import { DendronEngine } from "@dendronhq/engine-server";
import _ from "lodash";
import { URI } from "vscode-uri";
import { PodKind } from "./types";
import { Logger, createLogger } from "@dendronhq/common-server";

export type PodOptsV2 = {
  roots: string[];
  engine?: DEngine;
};

export type PodConfigEntry = {
  key: string;
  description: string;
  type: "string" | "number";
  default?: any;
};

export abstract class PodBaseV2 {
  public opts: PodOptsV2;
  private _engine: DEngine;
  public L: Logger;

  constructor(opts: PodOptsV2) {
    this.opts = opts;
    // @ts-ignore
    // TODO: change
    this.L = createLogger("PodLogger");
    if (!_.isUndefined(this.opts.engine)) {
      this._engine = this.opts.engine;
    } else {
      this._engine = DendronEngine.getOrCreateEngine({
        root: this.opts.roots[0],
      });
    }
  }

  get engine(): DEngine {
    return this._engine;
  }

  async initEngine() {
    await this.engine.init();
  }
}

export abstract class ExportPodBaseV2<
  TConfig extends ExportConfig = ExportConfig
> extends PodBaseV2 implements ExportPod<TConfig> {
  static kind = "export" as PodKind;

  cleanConfig(config: ExportConfig) {
    return {
      ..._.defaults(config, { includeStubs: false, includeBody: true }),
      dest: URI.file(config.dest),
    };
  }

  prepareForExport(opts: ExportPodOpts<TConfig>) {
    this.initEngine();
    let nodes = _.values(this.engine[opts.mode]);
    const hideBody = opts.config.includeBody ? false : true;
    if (!opts.config.includeStubs) {
      nodes = _.reject(nodes, { stub: true });
    }
    const payload: any = _.map(nodes, (ent) => {
      return ent.toRawProps(hideBody);
    });
    return payload;
  }

  abstract plant(opts: ExportPodOpts<TConfig>): Promise<void>;
}

export abstract class ImportPodBaseV2<TConfig extends ImportConfig = any>
  extends PodBaseV2
  implements ImportPod<TConfig> {
  static kind = "import" as PodKind;

  cleanConfig(config: ImportConfig) {
    return {
      ..._.defaults(config),
      src: URI.file(config.src),
    };
  }

  async prepare(_opts: ImportPodOpts<TConfig>) {
    await this.initEngine();
  }

  abstract plant(opts: ImportPodOpts<TConfig>): Promise<void>;
}

export type ExportConfig = {
  dest: string;
  /**
   * Default: false
   */
  includeStubs?: boolean;
  /**
   * Default: true
   */
  includeBody?: boolean;
};
export type ImportConfig = {
  src: string;
};

export type ExportPodOpts<TConfig extends ExportConfig> = {
  /**
   * Export schema or notes
   */
  mode: "notes" | "schemas";
  /**
   * Only export metadata?
   */
  config: TConfig;
};
export type ImportPodOpts<TConfig extends ImportConfig> = {
  /**
   * Export schema or notes
   */
  mode: "notes" | "schemas";
  /**
   * Only export metadata?
   */
  config: TConfig;
};

export interface ExportPod<TConfig extends ExportConfig = any> {
  plant: (opts: ExportPodOpts<TConfig>) => Promise<void>;
}

export type Pod = ExportPod;

export interface ImportPod<TConfig extends ImportConfig = any> {
  plant: (opts: ImportPodOpts<TConfig>) => Promise<void>;
}
