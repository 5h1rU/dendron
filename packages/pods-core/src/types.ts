import { DPod } from "./basev2";

export interface PodClassEntryV4 {
  id: string;
  description: string;
  kind: PodKind;
  new (): DPod<any>;
}

export type PodItemV4 = {
  id: string;
  description: string;
  podClass: PodClassEntryV4;
};

export type PodKind = "import" | "export" | "publish";
