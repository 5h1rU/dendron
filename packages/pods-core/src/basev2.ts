import { DEngineClientV2, DVault } from "@dendronhq/common-all";
import { createLogger, DLogger } from "@dendronhq/common-server";
import { URI } from "vscode-uri";

export abstract class BasePod {
  public L: DLogger;
  constructor() {
    this.L = createLogger("PodLogger");
  }
}

export type PodConfig = {
  key: string;
  description: string;
  type: "string" | "number" | "boolean";
  default?: any;
};

export type ExportPodRawConfig = {
  dest: string;
};

export type ExportPodCleanConfig = {
  dest: URI;
};

export type ExportPodPlantOpts<TConfig> = {
  wsRoot: string;
  vaults: DVault[];
  engine: DEngineClientV2;
  config: TConfig;
};

export type ExportPodExecuteOpts<TConfig extends ExportPodRawConfig> = {
  config: TConfig;
  engine: DEngineClientV2;
  wsRoot: string;
  vaults: DVault[];
};

export abstract class ExportPod<
  TConfigRaw extends ExportPodRawConfig,
  TConfigClean extends ExportPodCleanConfig
> {
  abstract config: PodConfig[];
  abstract clean(opts: {
    config: Partial<TConfigRaw>;
    wsRoot: string;
  }): Promise<TConfigClean>;
  abstract plant(opts: ExportPodPlantOpts<TConfigClean>): Promise<any>;

  async execute(opts: ExportPodExecuteOpts<TConfigRaw>) {
    const { config, engine, wsRoot, vaults } = opts;
    await engine.init();
    const cleanConfig = await this.clean({ config, wsRoot });
    return this.plant({ config: cleanConfig, wsRoot, vaults, engine });
  }
}
