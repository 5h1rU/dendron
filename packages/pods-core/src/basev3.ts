import {
  DendronError,
  DEngineClientV2,
  DVault,
  NotePropsV2,
  NoteUtilsV2,
  PodConfig,
  VaultUtils,
  WorkspaceOpts,
} from "@dendronhq/common-all";
import { createLogger, DLogger, resolvePath } from "@dendronhq/common-server";
import _ from "lodash";
import { URI } from "vscode-uri";
import { PodKind } from "./types";

export type PodOpts<T> = {
  engine: DEngineClientV2;
  config: T;
} & WorkspaceOpts;

// === Publish Pod

export type PublishPodExecuteOpts<T extends PublishPodConfig = any> = PodOpts<
  T
>;

export type PublishPodPlantOpts<
  T extends PublishPodConfig = any
> = PublishPodExecuteOpts<T> & { note: NotePropsV2 };

export type PublishPodConfig = {
  /**
   *  Name of file to publish
   */
  fname: string;
  /**
   * Name of vault
   */
  vaultName: string;
  /**
   * Where to write output to
   */
  dest: string | "stdout";
};

export abstract class PublishPod<T extends PublishPodConfig = any> {
  static kind = "publish" as PodKind;

  get config(): PodConfig[] {
    return [
      {
        key: "fname",
        description: "name of src file",
        type: "string" as const,
      },
      {
        key: "vaultName",
        description: "name of src vault",
        type: "string" as const,
      },
      {
        key: "dest",
        description: "where to export to",
        type: "string" as const,
      },
    ];
  }

  async execute(opts: PublishPodExecuteOpts<T>) {
    const { config, engine } = opts;
    const { fname, vaultName } = config;
    if (_.isUndefined(vaultName)) {
      throw new DendronError({ msg: "no vaultName" });
    }
    if (_.isUndefined(fname)) {
      throw new DendronError({ msg: "no fname" });
    }

    const vault = VaultUtils.getVaultByName({
      vaults: engine.vaultsv3,
      vname: vaultName,
    });
    const note = NoteUtilsV2.getNoteByFnameV5({
      fname,
      notes: engine.notes,
      vault: vault!,
      wsRoot: engine.wsRoot,
    });
    if (!note) {
      throw Error("no note found");
    }
    return this.plant({ ...opts, note });
  }

  abstract plant(opts: PublishPodPlantOpts<T>): Promise<string>;
}

// === Import Pod

export type ImportPodConfig = {
  /**
   * Where to import from
   */
  src: string;
  /**
   * Name of vault
   */
  vaultName: string;
  concatenate?: boolean;
  destName?: string;
  ignore?: boolean;
};
export type ImportPodExecuteOpts<T extends ImportPodConfig = any> = PodOpts<T>;

export type ImportPodPlantOpts<T extends ImportPodConfig = any> = Omit<
  ImportPodExecuteOpts<T>,
  "src"
> & { src: URI; vault: DVault };

export abstract class ImportPod<T extends ImportPodConfig = any> {
  public L: DLogger;
  static kind = "import" as PodKind;
  get config(): PodConfig[] {
    return [
      {
        key: "src",
        description: "Where to import from",
        type: "string" as const,
        required: true,
      },
      {
        key: "vaultName",
        description: "name of vault to import into",
        type: "string" as const,
        required: true,
      },
      {
        key: "concatenate",
        description: "whether to concatenate everything into one note",
        type: "boolean",
      },
      {
        key: "ignore",
        description: "whether to ignore a given file",
        type: "boolean",
      },
      {
        key: "destName",
        description: "If concatenate is set, name of destination path",
        type: "string" as const,
      },
    ];
  }
  constructor() {
    this.L = createLogger("ImportPod");
  }

  async execute(opts: ImportPodExecuteOpts<T>) {
    const { config, engine } = opts;
    const { src, vaultName, concatenate } = _.defaults(config, {
      concatenate: false,
    });

    // validate config
    if (_.isUndefined(src)) {
      throw new DendronError({ msg: "no src specified" });
    }
    if (_.isUndefined(vaultName)) {
      throw new DendronError({ msg: "no vaultName specified" });
    }
    if (concatenate && _.isUndefined(config?.destName)) {
      throw new DendronError({
        msg: "destName must be specified if concatenate is enabled",
      });
    }

    const vault = VaultUtils.getVaultByName({
      vaults: engine.vaultsv3,
      vname: vaultName,
    });
    const srcURL = URI.file(resolvePath(src, engine.wsRoot));
    this.L.info({ msg: "pre:plant", src: srcURL.fsPath });
    const out = await this.plant({ ...opts, src: srcURL, vault });
    this.L.info({ msg: "post:plant", src: srcURL.fsPath });
    return out;
  }
  abstract plant(opts: ImportPodPlantOpts<T>): Promise<NotePropsV2[]>;
}
