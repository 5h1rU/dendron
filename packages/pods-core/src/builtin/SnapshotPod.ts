import {
  ExportCleanConfig,
  ExportConfig,
  ExportPod,
  ExportPodBaseV3,
  ExportPodOpts,
  PodConfigEntry,
} from "../base";
import fs, { pathExistsSync } from "fs-extra";
import path from "path";
import { DVault, Time } from "@dendronhq/common-all";
import { URI } from "vscode-uri";
import _ from "lodash";
import { DUtils } from "@dendronhq/common-all";

const ID = "dendron.snapshot";

function genVaultId(vaultPath: string) {
  return path.basename(vaultPath);
}

type SnapshotExportPodConfig = ExportConfig & {
  ignore?: string;
};
type SnapshotExportPodCleanConfig = ExportCleanConfig & {
  ignore: string[];
};
type SnapshotExportPodOpts = ExportPodOpts<SnapshotExportPodConfig>;

export class SnapshotExportPod extends ExportPodBaseV3 {
  static id: string = ID;
  static description: string = "export notes to snapshot";

  static config = (): PodConfigEntry[] => {
    return [
      {
        key: "dest",
        description: "where will output be stored",
        type: "string",
        default: "$wsRoot/snapshots",
      },
      {
        key: "ignore",
        description: "what files will be ignored during snapshot",
        type: "string",
        default: ".git",
      },
    ];
  };

  async backupVault({
    vault,
    snapshotDirPath,
    ignore,
  }: {
    vault: DVault;
    snapshotDirPath: string;
    ignore: string[];
  }) {
    const vaultId = genVaultId(vault.fsPath);
    return fs.copy(vault.fsPath, path.join(snapshotDirPath, vaultId), {
      filter: (src) => {
        if (_.isEmpty(ignore)) {
          return true;
        }
        src = _.trimStart(_.replace(src, vault.fsPath, ""), "/");
        return !_.some(ignore, (ent) => {
          return DUtils.minimatch(src, ent);
        });
      },
    });
  }

  cleanConfig(config: SnapshotExportPodConfig): SnapshotExportPodCleanConfig {
    const cleanConfig = _.defaults(super.cleanConfig(config), {
      ignore: ".git",
    }) as any;
    cleanConfig.ignore = _.reject(cleanConfig.ignore.split(","), (ent) =>
      _.isEmpty(ent)
    );
    return cleanConfig;
  }

  async plant(opts: SnapshotExportPodOpts): Promise<any> {
    await this.initEngine();
    const cleanConfig = this.cleanConfig(opts.config);
    const { ignore } = cleanConfig;
    // const payload = this.prepareForExport(opts);

    // verify snapshot root
    const snapshotRoot = cleanConfig.dest.fsPath;
    fs.ensureDirSync(snapshotRoot);

    // create snapshot folder
    const snapshotDirId = Time.now().toMillis().toString();
    const snapshotDirPath = path.join(snapshotRoot, snapshotDirId);
    fs.ensureDirSync(snapshotDirPath);

    await Promise.all(
      this.opts.vaults.map((ent) => {
        return this.backupVault({
          vault: { fsPath: ent },
          snapshotDirPath,
          ignore,
        });
      })
    );
    return { snapshotDirPath };
  }
}
