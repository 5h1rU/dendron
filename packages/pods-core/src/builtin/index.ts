export * from "./JSONPod";
export * from "./FilePod";
