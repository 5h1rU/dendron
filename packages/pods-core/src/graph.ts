
type GraphCSSItem = {
  selector: string,
  property: string,
}

type GraphCSS = {
  obsidian: GraphCSSItem;
  dendron: GraphCSSItem;
}[]

const graphCSS: GraphCSS = [
  {
    obsidian: {
      selector: '.graph-view.color-fill',
      property: 'color',
    },
    dendron: {
      selector: 'node',
      property: 'background-color'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-fill',
      property: 'opacity',
    },
    dendron: {
      selector: 'node',
      property: 'background-opacity'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-circle',
      property: 'color',
    },
    dendron: {
      selector: 'node',
      property: 'border-color'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-line',
      property: 'color',
    },
    dendron: {
      selector: 'edge',
      property: 'line-color'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-line',
      property: 'opacity',
    },
    dendron: {
      selector: 'edge',
      property: 'line-opacity'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-text',
      property: 'color',
    },
    dendron: {
      selector: 'node',
      property: 'color'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-text',
      property: 'opacity',
    },
    dendron: {
      selector: 'node',
      property: 'text-opacity'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-fill-highlight',
      property: 'color',
    },
    dendron: {
      selector: 'node:selected',
      property: 'background-color'
    }
  },
  {
    obsidian: {
      selector: '.graph-view.color-fill-highlight',
      property: 'opacity',
    },
    dendron: {
      selector: 'node:selected',
      property: 'background-opacity'
    }
  },
]

export default graphCSS;