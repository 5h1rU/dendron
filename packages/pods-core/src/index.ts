import { readYAML } from "@dendronhq/common-server";
import fs, { ensureDirSync, writeFileSync } from "fs-extra";
import _ from "lodash";
import path from "path";
import { JSONExportPod, JSONImportPod, JSONPublishPod } from "./builtin";
import { MarkdownImportPod, MarkdownPublishPod } from "./builtin/MarkdownPod";
import { PodClassEntryV4 } from "./types";
export * from "./base";
export * from "./builtin";
export * from "./types";
export * from "./utils";

export function getAllExportPods(): PodClassEntryV4[] {
  return [JSONExportPod];
}
export function getAllPublishPods(): PodClassEntryV4[] {
  return [JSONPublishPod, MarkdownPublishPod];
}
export function getAllImportPods(): PodClassEntryV4[] {
  return [JSONImportPod, MarkdownImportPod];
}

// === utils

export class PodUtils {
  static getConfig({
    podsDir,
    podClass,
  }: {
    podsDir: string;
    podClass: PodClassEntryV4;
  }): false | any {
    const podConfigPath = PodUtils.getConfigPath({ podsDir, podClass });
    if (!fs.existsSync(podConfigPath)) {
      return false;
    } else {
      return readYAML(podConfigPath);
    }
  }

  static getConfigPath({
    podsDir,
    podClass,
  }: {
    podsDir: string;
    podClass: PodClassEntryV4;
  }): string {
    return path.join(podsDir, podClass.id, `config.${podClass.kind}.yml`);
  }

  static getPath({
    podsDir,
    podClass,
  }: {
    podsDir: string;
    podClass: PodClassEntryV4;
  }): string {
    return path.join(podsDir, podClass.id);
  }

  static genConfigFile({
    podsDir,
    podClass,
  }: {
    podsDir: string;
    podClass: PodClassEntryV4;
  }) {
    const podConfigPath = PodUtils.getConfigPath({ podsDir, podClass });
    ensureDirSync(path.dirname(podConfigPath));
    const pod = new podClass();
    const config = pod.config
      .map((ent) => {
        ent = _.defaults(ent, { default: "TODO" });
        return [
          `# description: ${ent.description}`,
          `# type: ${ent.type}`,
          `${ent.key}: ${ent.default}`,
        ].join("\n");
      })
      .join("\n");
    if (!fs.existsSync(podConfigPath)) {
      writeFileSync(podConfigPath, config);
    }
    return podConfigPath;
  }

  static hasRequiredOpts(_pClassEntry: PodClassEntryV4): boolean {
    // TODO:
    return false;
  }
}
