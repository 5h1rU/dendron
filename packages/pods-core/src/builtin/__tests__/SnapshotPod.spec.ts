import { DEngineClientV2 } from "@dendronhq/common-all/src";
import { createLogger } from "@dendronhq/common-server/src";
import {
  EngineTestUtilsV2,
  NodeTestPresetsV2,
} from "@dendronhq/common-test-utils";
import { DendronEngineV2, FileStorageV2 } from "@dendronhq/engine-server";
import fs from "fs-extra";
import path from "path";
import { SnapshotExportPod } from "../SnapshotPod";

describe("SnapshotPod", () => {
  let vaults: string[];
  let wsRoot: string;
  let engine: DEngineClientV2;

  beforeEach(async () => {
    ({ wsRoot, vaults } = await EngineTestUtilsV2.setupWS({
      initDirCb: async (vaultDir) => {
        await NodeTestPresetsV2.createOneNoteOneSchemaPresetWithBody({
          vaultDir,
        });
      },
    }));
    const LOGGER = createLogger();
    engine = new DendronEngineV2({
      vaults,
      forceNew: true,
      store: new FileStorageV2({ vaults, logger: LOGGER }),
      mode: "fuzzy",
      logger: LOGGER,
    });
  });

  test("basic", async () => {
    let dest = path.join(wsRoot, "snapshot");
    const pod = new SnapshotExportPod({ vaults, wsRoot, engine });
    fs.ensureDirSync(dest);
    const { snapshotDirPath } = await pod.plant({
      config: {
        dest,
      },
      mode: "notes",
    });
    const snapshotDir = fs.readdirSync(snapshotDirPath);

    expect(snapshotDir).toMatchSnapshot();
    // backup vault
    const vaultPath = path.join(snapshotDirPath, "vault");
    const snapshotVault = fs.readdirSync(vaultPath);
    expect(snapshotVault).toMatchSnapshot();
    expect(snapshotVault.length).toEqual(6);
    // copy assets
    const assetsDir = fs.readdirSync(path.join(vaultPath, "assets"));
    expect(assetsDir).toEqual(["foo.jpg"]);
  });

  test("empty ignore", async () => {
    let dest = path.join(wsRoot, "snapshot");
    const pod = new SnapshotExportPod({ vaults, wsRoot, engine });
    fs.ensureDirSync(dest);
    const { snapshotDirPath } = await pod.plant({
      config: {
        dest,
        ignore: "",
      },
      mode: "notes",
    });
    const snapshotDir = fs.readdirSync(snapshotDirPath);

    expect(snapshotDir).toMatchSnapshot();
    // backup vault
    const vaultPath = path.join(snapshotDirPath, "vault");
    const snapshotVault = fs.readdirSync(vaultPath);
    expect(snapshotVault).toMatchSnapshot();
    expect(snapshotVault.length).toEqual(7);
    // copy assets
    const assetsDir = fs.readdirSync(path.join(vaultPath, "assets"));
    expect(assetsDir).toEqual(["foo.jpg"]);
  });

  test("ignore foo", async () => {
    let dest = path.join(wsRoot, "snapshot");
    const pod = new SnapshotExportPod({ vaults, wsRoot, engine });
    fs.ensureDirSync(dest);
    const { snapshotDirPath } = await pod.plant({
      config: {
        dest,
        ignore: "foo*",
      },
      mode: "notes",
    });
    const snapshotDir = fs.readdirSync(snapshotDirPath);

    expect(snapshotDir).toMatchSnapshot();
    // backup vault
    const vaultPath = path.join(snapshotDirPath, "vault");
    const snapshotVault = fs.readdirSync(vaultPath);
    expect(snapshotVault).toMatchSnapshot();
    expect(snapshotVault.length).toEqual(4);
    // copy assets
    const assetsDir = fs.readdirSync(path.join(vaultPath, "assets"));
    expect(assetsDir).toEqual(["foo.jpg"]);
  });
});
