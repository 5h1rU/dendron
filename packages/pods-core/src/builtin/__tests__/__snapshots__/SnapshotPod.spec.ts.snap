// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapshotPod basic 1`] = `
Array [
  "vault",
]
`;

exports[`SnapshotPod basic 2`] = `
Array [
  "assets",
  "foo.ch1.md",
  "foo.md",
  "foo.schema.yml",
  "root.md",
  "root.schema.yml",
]
`;

exports[`SnapshotPod empty ignore 1`] = `
Array [
  "vault",
]
`;

exports[`SnapshotPod empty ignore 2`] = `
Array [
  ".git",
  "assets",
  "foo.ch1.md",
  "foo.md",
  "foo.schema.yml",
  "root.md",
  "root.schema.yml",
]
`;

exports[`SnapshotPod ignore foo 1`] = `
Array [
  "vault",
]
`;

exports[`SnapshotPod ignore foo 2`] = `
Array [
  ".git",
  "assets",
  "root.md",
  "root.schema.yml",
]
`;
